sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/regexpExpressions.scala

```diff
@@ -681,7 +682,7 @@ case class RegExpReplace(subject: Expression, regexp: Expression, rep: Expressio
   @transient private var lastReplacement: String = _
   @transient private var lastReplacementInUTF8: UTF8String = _
   // result buffer write by Matcher
-  @transient private lazy val result: StringBuffer = new StringBuffer
+  @transient private lazy val result: JStringBuilder = new JStringBuilder
```

Comment:
Why do we not need to use thread safe StringBuffer? Will this code be never run in a multithreaded environment?

Reply:
Because in Spark, `each task` is an independent thread.

Reply:
Oh okay! There are some other places in the codebase where we use StringBuffer. We may be able to use StringBuilder there and see some performance gain.

Reply:
I noticed another function `FormatNumber`, but according to the results of my local benchmark test, I did not see great performance gain (`almost no change`).
https://github.com/apache/spark/blob/beea9c36725e7aa8f17f952da70e1d25013dc336/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala#L3345

```scala
object FormatNumberBenchmark extends SqlBasedBenchmark {
  private val N = 1000000 // 10_000_00
  private val M = 100

  private val df = spark.range(N).to(new StructType().add("id", "int")).
    withColumn("format1", lit("##################.##")).
    withColumn("format2", lit(4)).
    withColumn("value", col("id") / M)

  private def doBenchmark(): Unit = {
    // df.selectExpr("format_number(value, format2)", "format_number(value, format1)").noop()
    df.selectExpr("format_number(value, format2)").noop()
  }

  override def runBenchmarkSuite(mainArgs: Array[String]): Unit = {
    runBenchmark("format_number") {
      val benchmark = new Benchmark("format_number", N, output = output)
      benchmark.addCase("optimize", M) { _ =>
        doBenchmark()
      }
      benchmark.run()
    }
  }
}

```

Reply:
All right! Thank you so much